clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -136,27 +369,160 @@ public ConfigDef withClientSaslSupport() {
      * the appropriate type (int, string, etc)
      */
     public Map<String, Object> parse(Map<?, ?> props) {
+        return parse(props, true);
+    }
+
+    /**
+     * Parse and validate configs against this configuration definition. The input is a map of configs. It is expected
+     * that the keys of the map are strings, but the values can either be strings or they may already be of the
+     * appropriate type (int, string, etc). This will work equally well with either java.util.Properties instances or a
+     * programmatically constructed map.
+     *
+     * @param props The configs to parse and validate
+     * @param checkRequired Whether to throw a config exception in case that we don't provide a value for a required config
+     * @return Parsed and validated configs. The key will be the config name and the value will be the value parsed into
+     * the appropriate type (int, string, etc)
+     */
+    public Map<String, Object> parse(Map<?, ?> props, boolean checkRequired) {
+        ensureDependentsDefined();
         /* parse all known keys */
-        Map<String, Object> values = new HashMap<String, Object>();
+        Map<String, Object> values = new HashMap<>();
         for (ConfigKey key : configKeys.values()) {
-            Object value;
+            Object value = null;
             // props map contains setting - assign ConfigKey value
-            if (props.containsKey(key.name))
+            if (props.containsKey(key.name)) {
                 value = parseType(key.name, props.get(key.name), key.type);
-            // props map doesn't contain setting, the key is required because no default value specified - its an error
-            else if (key.defaultValue == NO_DEFAULT_VALUE)
-                throw new ConfigException("Missing required configuration \"" + key.name + "\" which has no default value.");
-            // otherwise assign setting its default value
-            else
+                // props map doesn't contain setting, the key is required because no default value specified - its an error
+            } else if (key.defaultValue == NO_DEFAULT_VALUE) {
+                if (checkRequired) {
+                    throw new ConfigException("Missing required configuration \"" + key.name + "\" which has no default value.");
+                }
+            } else {
+                // otherwise assign setting its default value
                 value = key.defaultValue;
-            if (key.validator != null)
+            }
+            if (value != null && key.validator != null) {
                 key.validator.ensureValid(key.name, value);
```

Comment:
Same question here about `checkRequired`. Maybe a different name is needed, something like `allowErrors` and the entire body of the `for` loop could be protected by a `try`/`catch`?